### PR TITLE
Create TestClient

### PR DESCRIPTION
**Pull request by swoogles**
The WebSocket behavior is not implemented yet. 
It will require some changes to `Channel` that will be more invasive than this PR.

### CODE REVIEW

File: zio-http-testkit/src/main/scala/zio/http/TestClient.scala [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  ): ZIO[R, Nothing, Unit] =
+    for {
+      r                <- ZIO.environment[R]
+      previousBehavior <- behavior.get
+      newBehavior                  = handler.andThen(_.provideEnvironment(r))
+      app: HttpApp[Any, Throwable] = Http.collectZIO(newBehavior)
+      _ <- behavior.set(previousBehavior ++ app)
+    } yield ()
+
+  val headers: Headers                   = Headers.empty
```
Comment (jdegoes) [inline zio-http-testkit/src/main/scala/zio/http/TestClient.scala:69]: What are these for?

File: zio-http-testkit/src/main/scala/zio/http/TestClient.scala [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import zio.http.socket.SocketApp
+
+/**
+ * Enables tests that use a client without needing a live Server
+ *
+ * @param behavior
+ *   Contains the user-specified behavior that takes the place of the usual
+ *   Server
+ */
+case class TestClient(behavior: Ref[HttpApp[Any, Throwable]]) extends Client {
```
Comment (jdegoes) [inline zio-http-testkit/src/main/scala/zio/http/TestClient.scala]:
```suggestion
final case class TestClient(behavior: Ref[HttpApp[Any, Throwable]]) extends Client {
```

File: zio-http-testkit/src/main/scala/zio/http/TestClient.scala [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+   *   {{{
+   *    TestClient.addHandler{case request  if request.method == Method.GET => ZIO.succeed(Response.ok)}
+   *   }}}
+   */
+  def addHandler[R](
+    handler: PartialFunction[Request, ZIO[R, Throwable, Response]],
+  ): ZIO[R with TestClient, Nothing, Unit] =
+    ZIO.serviceWithZIO[TestClient](_.addHandler(handler))
+
+  val layer: ZLayer[Any, Throwable, TestClient] =
```
Comment (jdegoes) [inline zio-http-testkit/src/main/scala/zio/http/TestClient.scala]: The error is `Throwable` but there is no way for this layer to fail?

File: zio-http-testkit/src/main/scala/zio/http/TestClient.scala (hunk 1 above)
Comment (jdegoes) [reply zio-http-testkit/src/main/scala/zio/http/TestClient.scala:69]: Ah they come from `Client`

File: zio-http-testkit/src/main/scala/zio/http/TestClient.scala (hunk 3 above)
Comment (swoogles) [reply zio-http-testkit/src/main/scala/zio/http/TestClient.scala]: Thanks, removed!

**Review by jdegoes — APPROVED** [review]

**Pull request merged**